kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMetadata.java

```diff
@@ -434,14 +437,14 @@ public TxnTransitMetadata prepareAddPartitions(Set<TopicPartition> addedTopicPar
         if (topicPartitions != null) {
             newPartitionSet.addAll(topicPartitions);
         }
-        newPartitionSet.addAll(new HashSet<>(addedTopicPartitions));
+        newPartitionSet.addAll(addedTopicPartitions);
         return prepareTransitionTo(TransactionState.ONGOING, producerId, producerEpoch, lastProducerEpoch,
-                txnTimeoutMs, newPartitionSet, newTxnStartTimestamp, updateTimestamp);
+                txnTimeoutMs, ImmutableSet.copyOf(newPartitionSet), newTxnStartTimestamp, updateTimestamp);
```

Comment:
Here we don't need to copy again. You can just use `ImmutableSet.copyOf((topicPartitions != null) ? topicPartitions : ImmutableSet.of());`

Reply:
Here we also added all `addedTopicPartitions ` into `newPartitionSet`.

Reply:
Oh I missed it. Then we can replace it with:

```java
        ImmutableSet<TopicPartition> partitions = ImmutableSet.<TopicPartition>builder()
                .addAll((topicPartitions != null) ? topicPartitions : ImmutableSet.of())
                .addAll(addedTopicPartitions)
                .build();
```

Reply:
Fixed.